keps/sig-node/3545-improved-multi-numa-alignment/README.md

```diff
@@ -0,0 +1,444 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+We propose an enhancement to TopologyManager that enables it to favor sets of NUMA nodes with shorter distance between the nodes, when making admission decision. The proposed enhancement is only applicable when comparing sets of NUMA nodes equal in size.
+
+## Motivation
+
+To support latency-critical execution and high-throughput applications (including those running on Kubernetes), NUMA locality(topology) of resources assigned to the workloads is crucial. THe best NUMA locality can be achieved by minimizing number of NUMA nodes required to execute the workload, and minimizing the NUMA distance between the nodes. NUMA distance is a metric which exposes relative physical distance between NUMA cells which relates to latency that is introduced when a CPU needs to access resources from different NUMA nodes. The distance is always within range 10-254 and for a local access its value is 10.
+
+As of now, TopologyManager provides policies to either provide resources from only single numa node or the smallest number of NUMA nodes. TopologyManager is not aware of the NUMA distances and does not take it into consideration during admission decisions. 
+
+This limitation surfaces in multi socket, as well as single socket multi NUMA systems (commonly present in AMD hardware), and might cause a significant performance degradation for latency-critical execution and high-throughput applications, in a case when the Topology Manager schedules a pod on non-adjacent numa nodes.
+
+
+### Goals
+
+* Enable TopologyManager to prefer sets of NUMA nodes with shorter NUMA distances for all TopologyManager policies
+* Preserve all other properties of all policies
+
+## Proposal
+
+### Risks and Mitigations
+
+| Risk                                             | Impact | Mitigation                                                                                                                                                                               |
+|--------------------------------------------------|--------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Bugs in the implementation lead to kubelet crash | High   | Disable the policy option and restart the kubelet. The workload will run but resources allocation can be spread among NUMA nodes which are non-adjacent what can lead to higher latency. |
+
+## Design Details
+
+
+### Proposed Change
+ 
+We propose to
+- add a new flag in Kubelet called `TopologyManagerPolicyOptions` in the kubelet config or command line argument called `topology-manager-policy-options` which allows the user to specify the Topology Manager policy option.
```

Comment:
IIUC we want to replicate the policy options management we added in https://github.com/kubernetes/enhancements/issues/2625 , right?  Doing this for the cpumanager policy options required nontrivial discussion and the outcome was like the first of a kind for k8s. I for myself think this process is sustainable enough and makes sense, but we need to doublecheck we can replicate with topology manager soliciting feedback. I think this will emerge during the review of this KEP, let's make sure it happens though.

Reply:
You are right, we wanted to replicate the policy options in CPUManager to be consistent how we handle new policies/options regarding resource management in Kubelet.

Reply:
My hope is that it will be easier to digest this time around since there is already precedence. We should definitely pull in the same people that raised questions last time though. Can you /cc them here?

Reply:
/cc @derekwaynecarr

Reply:
I'll look to the discussion happened on sig-arch and CC the relevant folks.

Reply:
back in time we discussed with prod-readiness and in sig-arch with @johnbelamaric

Reply:
Agreed. Since there actually is precedent now, I don't think it should be a problem to follow the same process, but we should get confirmation.

Reply:
FWIW this seems reasonable to me but I will defer to those involved last time